oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java

```diff
@@ -0,0 +1,455 @@
+    checkNotNull(json);
+    checkNotNull(transportFactory);
+
+    String audience = (String) json.get("audience");
+    String subjectTokenType = (String) json.get("subject_token_type");
+    String tokenUrl = (String) json.get("token_url");
+    String serviceAccountImpersonationUrl = (String) json.get("service_account_impersonation_url");
+
+    Map<String, Object> credentialSourceMap = (Map<String, Object>) json.get("credential_source");
+
+    // Optional params.
+    String tokenInfoUrl =
+        json.containsKey("token_info_url") ? (String) json.get("token_info_url") : null;
+    String clientId = json.containsKey("client_id") ? (String) json.get("client_id") : null;
+    String clientSecret =
+        json.containsKey("client_secret") ? (String) json.get("client_secret") : null;
+    String quotaProjectId =
+        json.containsKey("quota_project_id") ? (String) json.get("quota_project_id") : null;
+
+    if (isAwsCredential(credentialSourceMap)) {
+      return new AwsCredentials(
+          transportFactory,
+          audience,
+          subjectTokenType,
+          tokenUrl,
+          new AwsCredentialSource(credentialSourceMap),
+          tokenInfoUrl,
+          serviceAccountImpersonationUrl,
+          quotaProjectId,
+          clientId,
+          clientSecret,
+          /* scopes= */ null);
+    }
+    return new IdentityPoolCredentials(
+        transportFactory,
+        audience,
+        subjectTokenType,
+        tokenUrl,
+        new IdentityPoolCredentialSource(credentialSourceMap),
+        tokenInfoUrl,
+        serviceAccountImpersonationUrl,
+        quotaProjectId,
+        clientId,
+        clientSecret,
+        /* scopes= */ null);
+  }
+
+  private static boolean isAwsCredential(Map<String, Object> credentialSource) {
+    return credentialSource.containsKey("environment_id")
+        && ((String) credentialSource.get("environment_id")).startsWith("aws");
+  }
+
+  /**
+   * Exchanges the external credential for a GCP access token.
+   *
+   * @param stsTokenExchangeRequest the STS token exchange request
+   * @return the access token returned by STS
+   * @throws OAuthException if the call to STS fails
+   */
+  protected AccessToken exchangeExternalCredentialForAccessToken(
+      StsTokenExchangeRequest stsTokenExchangeRequest) throws IOException {
+    // Handle service account impersonation if necessary.
+    if (impersonatedCredentials != null) {
+      return impersonatedCredentials.refreshAccessToken();
+    }
+
+    StsRequestHandler requestHandler =
+        StsRequestHandler.newBuilder(
+                tokenUrl, stsTokenExchangeRequest, transportFactory.create().createRequestFactory())
+            .build();
+
+    StsTokenExchangeResponse response = requestHandler.exchangeToken();
+    return response.getAccessToken();
+  }
+
+  private static String extractTargetPrincipal(String serviceAccountImpersonationUrl) {
+    // Extract the target principle.
```

Comment:
principle or principal? I can never remember which is which, but either the method name or the comment is incorrect.